File: ts/parser/src/blocks.ts
// Line-oriented block parser: a faithful port of rust/parser/src/blocks.rs.
// Where this file and that file disagree, one of them has a conformance bug.

import type { Node, Reason } from "./ast.ts";
import { isName, nameLen, parseAttrs, utf8Len } from "./attrs.ts";
import { parseInlines } from "./inlines.ts";

export const MAX_LIST_DEPTH = 8;
export const MAX_QUOTE_DEPTH = 8;
export const MAX_DIRECTIVE_DEPTH = 4;
export const MAX_TARGET_BYTES = 2048;

interface Ctx {
  dirDepth: number;
  listDepth: number;
  quoteDepth: number;
}

interface Cursor {
  lines: string[];
  pos: number;
}

function peek(cur: Cursor): string | undefined {
  return cur.lines[cur.pos];
}

export function parseBlocks(body: string): Node[] {
  const lines = body.split("\n");
  // A final newline terminates the last line rather than creating an empty
  // one (otherwise every trailing newline leaks a blank into fences).
  if (lines.length > 0 && lines[lines.length - 1] === "") {
    lines.pop();
  }
  const cur: Cursor = { lines, pos: 0 };
  return parseContainer(cur, { dirDepth: 0, listDepth: 0, quoteDepth: 0 }, null);
}

/**
 * Parse blocks until the container ends. `openDir` is the name of the
 * directive whose body this is (null at document level and inside
 * blockquotes / list items, which are fresh containers).
 */
function parseContainer(cur: Cursor, ctx: Ctx, openDir: string | null): Node[] {
  const out: Node[] = [];
  for (;;) {
    const line = peek(cur);
    if (line === undefined) {
      break;
    }
    if (isBlank(line)) {
      cur.pos += 1;
      continue;
    }

    // Directive open / close / error lines.
    if (line.startsWith(":::")) {
      cur.pos += 1;
      const dir = classifyDirective(line);
      if (dir.kind === "close") {
        if (openDir !== null && dir.name === null) {
          return out;
        }
        if (openDir !== null && dir.name === openDir) {
          return out;
        }
        out.push(invalid(openDir !== null ? "mismatched_close" : "stray_close"));
      } else if (dir.kind === "open") {
        if (ctx.dirDepth >= MAX_DIRECTIVE_DEPTH) {
          out.push(invalid("depth_exceeded"));
          continue;
        }
        const attrs = parseAttrs(dir.attrsSrc);
        if (!attrs.ok) {
          out.push(invalid(attrs.reason));
        } else {
          const children = dir.leaf
            ? []
            : parseContainer(cur, { ...ctx, dirDepth: ctx.dirDepth + 1 }, dir.name);
          out.push({ type: "directive", name: dir.name, attrs: attrs.attrs, children });
        }
      } else {
        out.push(invalid(dir.reason));
      }
      continue;
    }

    // Comment block: consecutive `%%` lines, raw content.
    if (line.startsWith("%%")) {
      const texts: string[] = [];
      for (let l = peek(cur); l !== undefined && l.startsWith("%%"); l = peek(cur)) {
        const rest = l.slice(2);
        texts.push(rest.startsWith(" ") ? rest.slice(1) : rest);
        cur.pos += 1;
      }
      out.push({ type: "comment", text: texts.join("\n") });
      continue;
    }

    const heading = headingLine(line);
    if (heading !== null) {
      cur.pos += 1;
      out.push({
        type: "heading",
        level: heading.level,
        children: parseInlines(trimSpaceTab(heading.content)),
      });
      continue;
    }

    const fence = fenceOpen(line);
    if (fence !== null) {
      cur.pos += 1;
      const content: string[] = [];
      for (let l = peek(cur); l !== undefined; l = peek(cur)) {
        cur.pos += 1;
        if (fenceClose(l, fence.len)) {
          break;
        }
        content.push(l);
      }
      const text = content.join("\n");
      out.push(
        fence.info === null
          ? { type: "code_block", text }
          : { type: "code_block", info: fence.info, text },
      );
      continue;
    }

    if (line.startsWith(">") && ctx.quoteDepth < MAX_QUOTE_DEPTH) {
      const innerLines: string[] = [];
      for (let l = peek(cur); l !== undefined && l.startsWith(">"); l = peek(cur)) {
        const rest = l.slice(1);
        innerLines.push(rest.startsWith(" ") ? rest.slice(1) : rest);
        cur.pos += 1;
      }
      const sub: Cursor = { lines: innerLines, pos: 0 };
      out.push({
        type: "blockquote",
        children: parseContainer(sub, { ...ctx, quoteDepth: ctx.quoteDepth + 1 }, null),
      });
      continue;
    }

    if (line.replace(/[ \t]+$/, "") === "---") {
      cur.pos += 1;
      out.push({ type: "thematic_break" });
      continue;
    }

    if (ctx.listDepth < MAX_LIST_DEPTH) {
      const m = marker(line);
      if (m !== null) {
        out.push(parseList(cur, ctx, m));
        continue;
      }
    }

    // Paragraph: plain lines until blank / container end / block start.
    const plines = [line];
    cur.pos += 1;
    for (let l = peek(cur); l !== undefined; l = peek(cur)) {
      if (isBlank(l) || isBlockStart(l, ctx)) {
        break;
      }
      plines.push(l);
      cur.pos += 1;
    }
    if (plines.length === 1) {
      const target = turbolink(plines[0]!);
      if (target !== null) {
        out.push({ type: "turbolink", target });
        continue;
      }
    }
    out.push({ type: "paragraph", children: parseInlines(plines.join("\n")) });
  }
  return out;
}

function invalid(reason: Reason): Node {
  return { type: "invalid_directive", reason, children: [] };
}

function isBlank(line: string): boolean {
  return /^[ \t]*$/.test(line);
}

function indentOf(line: string): number {
  let i = 0;
  while (i < line.length && line[i] === " ") {
    i += 1;
  }
  return i;
}

function trimSpaceTab(s: string): string {
  return s.replace(/^[ \t]+|[ \t]+$/g, "");
}

/** Would this line start a non-paragraph block in this context? (Used both
 * by the dispatcher and to end paragraphs: any block construct interrupts.) */
function isBlockStart(line: string, ctx: Ctx): boolean {
  return (
    line.startsWith(":::") ||
    line.startsWith("%%") ||
    headingLine(line) !== null ||
    fenceOpen(line) !== null ||
    (line.startsWith(">") && ctx.quoteDepth < MAX_QUOTE_DEPTH) ||
    line.replace(/[ \t]+$/, "") === "---" ||
    (ctx.listDepth < MAX_LIST_DEPTH && marker(line) !== null)
  );
}

function headingLine(line: string): { level: number; content: string } | null {
  let n = 0;
  while (n < line.length && line[n] === "#") {
    n += 1;
  }
  if (n >= 1 && n <= 6 && line[n] === " ") {
    return { level: n, content: line.slice(n + 1) };
  }
  return null;
}

function fenceOpen(line: string): { len: number; info: string | null } | null {
  let n = 0;
  while (n < line.length && line[n] === "`") {
    n += 1;
  }
  if (n < 3) {
    return null;
  }
  const info = trimSpaceTab(line.slice(n));
  return { len: n, info: info === "" ? null : info };
}

function fenceClose(line: string, openLen: number): boolean {
  const t = line.replace(/[ \t]+$/, "");
  return t.length >= openLen && /^`+$/.test(t);
}

interface Marker {
  indent: number;
  ordered: boolean;
  contentIdx: number;
}

function marker(line: string): Marker | null {
  const indent = indentOf(line);
  const rest = line.slice(indent);
  if (rest.length >= 2 && "-*+".includes(rest[0]!) && rest[1] === " ") {
    return { indent, ordered: false, contentIdx: indent + 2 };
  }
  const m = /^([0-9]+)\. /.exec(rest);
  if (m !== null) {
    return { indent, ordered: true, contentIdx: indent + m[1]!.length + 2 };
  }
  return null;
}

/** One list of one kind. A same-column marker of the other kind ends this
 * list (the dispatcher immediately starts the next one). */
function parseList(cur: Cursor, ctx: Ctx, first: Marker): Node {
  const ordered = first.ordered;
  const items: Node[] = [];
  let buf: string[] = [peek(cur)!.slice(first.contentIdx)];
  cur.pos += 1;

  for (;;) {
    const line = peek(cur);
    if (line === undefined) {
      break;
    }
    if (isBlank(line)) {
      // The list continues past blanks only into indented content or a
      // same-kind column-0/1 marker; anything else ends it here.
      let j = cur.pos + 1;
      while (j < cur.lines.length && isBlank(cur.lines[j]!)) {
        j += 1;
      }
      const next = cur.lines[j];
      let continues = false;
      if (next !== undefined) {
        if (indentOf(next) >= 2) {
          continues = true;
        } else {
          const m = marker(next);
          continues = m !== null && m.ordered === ordered && m.indent < 2;
        }
      }
      if (!continues) {
        break;
      }
      buf.push("");
      cur.pos += 1;
      continue;
    }
    if (indentOf(line) >= 2) {
      // Content (or a deeper marker) inside the current item; strip the
      // content column. Off-grid extra spaces ride along (floor rule).
      buf.push(line.slice(2));
      cur.pos += 1;
      continue;
    }
    // Column 0 or 1 (floors to 0).
    const m = marker(line);
    if (m !== null && m.ordered === ordered) {
      items.push(finishItem(buf, ctx));
      buf = [line.slice(m.contentIdx)];
      cur.pos += 1;
      continue;
    }
    break; // column-0 block, or a kind switch
  }
  items.push(finishItem(buf, ctx));
  return { type: "list", ordered, children: items };
}

function finishItem(buf: string[], ctx: Ctx): Node {
  const sub: Cursor = { lines: buf, pos: 0 };
  return {
    type: "list_item",
    children: parseContainer(sub, { ...ctx, listDepth: ctx.listDepth + 1 }, null),
  };
}

type DirLine =
  | { kind: "open"; name: string; attrsSrc: string; leaf: boolean }
  | { kind: "close"; name: string | null }
  | { kind: "bad"; reason: Reason };

function classifyDirective(line: string): DirLine {
  const rest = line.slice(3).replace(/[ \t]+$/, "");
  if (rest === "") {
    return { kind: "close", name: null };
  }
  if (rest.startsWith(" ") || rest.startsWith("\t")) {
    // `::: name` - a named close.
    const name = trimSpaceTab(rest);
    return isName(name)
      ? { kind: "close", name }
      : { kind: "bad", reason: "bad_name" };
  }
  // Open line. The leaf closer token is stripped before attribute parsing.
  const leaf = rest.endsWith(":::");
  const body = leaf ? rest.slice(0, -3) : rest;
  const nlen = nameLen(body);
  if (nlen === 0) {
    return { kind: "bad", reason: "bad_name" };
  }
  const after = body.slice(nlen);
  if (!(after === "" || after.startsWith(" ") || after.startsWith("\t"))) {
    return { kind: "bad", reason: "bad_name" };
  }
  return { kind: "open", name: body.slice(0, nlen), attrsSrc: after, leaf };
}

/**
 * A paragraph that is exactly one authority-form absolute URI is a
 * turbolink. Every bare word is a valid relative URI reference, so the
 * sugar demands `scheme://`; everything else uses `:::turbolink`.
 */
function turbolink(line: string): string | null {
  const t = trimSpaceTab(line);
  if (t === "" || utf8Len(t) > MAX_TARGET_BYTES) {
    return null;
  }
  if (t.includes(" ") || t.includes("\t")) {
    return null;
  }
  const m = /^[A-Za-z][A-Za-z0-9+.-]*/.exec(t);
  if (m === null) {
    return null;
  }
  const after = t.slice(m[0].length);
  return after.startsWith("://") && after.length > 3 ? t : null;
}

File: rust/parser/src/blocks.rs
//! Line-oriented block parser. Bracket-LIFO directives, no per-element
//! implicit-close rules: recovery is trivial to specify and trivial to match.

use crate::ast::{Node, Reason};
use crate::attrs::{is_name, name_len, parse_attrs};
use crate::inlines::parse_inlines;

pub const MAX_LIST_DEPTH: usize = 16;
pub const MAX_QUOTE_DEPTH: usize = 16;
pub const MAX_DIRECTIVE_DEPTH: usize = 8;
pub const MAX_TARGET_BYTES: usize = 4096;

#[derive(Clone, Copy, Default)]
struct Ctx {
    dir_depth: usize,
    list_depth: usize,
    quote_depth: usize,
}

struct Cursor<'a> {
    lines: Vec<&'a str>,
    pos: usize,
}

impl<'a> Cursor<'a> {
    fn peek(&self) -> Option<&'a str> {
        self.lines.get(self.pos).copied()
    }
}

pub fn parse_blocks(body: &str) -> Vec<Node> {
    let mut lines: Vec<&str> = body.split('\n').collect();
    // A final newline terminates the last line rather than creating an empty
    // one (otherwise every trailing newline leaks a blank into fences).
    if lines.last() == Some(&"") {
        lines.pop();
    }
    let mut cur = Cursor { lines, pos: 0 };
    parse_container(&mut cur, Ctx::default(), None)
}

/// Parse blocks until the container ends. `open_dir` is the name of the
/// directive whose body this is (None at document level and inside
/// blockquotes / list items, which are fresh containers).
fn parse_container(cur: &mut Cursor, ctx: Ctx, open_dir: Option<&str>) -> Vec<Node> {
    let mut out: Vec<Node> = Vec::new();
    while let Some(line) = cur.peek() {
        if is_blank(line) {
            cur.pos += 1;
            continue;
        }

        // Directive open / close / error lines.
        if line.starts_with(":::") {
            cur.pos += 1;
            match classify_directive(line) {
                DirLine::Close(name) => match (open_dir, name) {
                    (Some(_), None) => return out,
                    (Some(open), Some(n)) if n == open => return out,
                    (Some(_), Some(_)) => out.push(invalid(Reason::MismatchedClose)),
                    (None, _) => out.push(invalid(Reason::StrayClose)),
                },
                DirLine::Open { name, attrs_src, leaf } => {
                    if ctx.dir_depth >= MAX_DIRECTIVE_DEPTH {
                        out.push(invalid(Reason::DepthExceeded));
                        continue;
                    }
                    match parse_attrs(&attrs_src) {
                        Err(reason) => out.push(invalid(reason)),
                        Ok(attrs) => {
                            let children = if leaf {
                                Vec::new()
                            } else {
                                let mut inner = ctx;
                                inner.dir_depth += 1;
                                parse_container(cur, inner, Some(&name))
                            };
                            out.push(Node::Directive { name, attrs, children });
                        }
                    }
                }
                DirLine::Bad(reason) => out.push(invalid(reason)),
            }
            continue;
        }

        // Comment block: consecutive `%%` lines, raw content.
        if line.starts_with("%%") {
            let mut texts: Vec<&str> = Vec::new();
            while let Some(l) = cur.peek() {
                match l.strip_prefix("%%") {
                    Some(rest) => {
                        texts.push(rest.strip_prefix(' ').unwrap_or(rest));
                        cur.pos += 1;
                    }
                    None => break,
                }
            }
            out.push(Node::Comment { text: texts.join("\n") });
            continue;
        }

        if let Some((level, content)) = heading_line(line) {
            cur.pos += 1;
            out.push(Node::Heading {
                level,
                children: parse_inlines(content.trim_matches([' ', '\t'])),
            });
            continue;
        }

        if let Some((fence_len, info)) = fence_open(line) {
            cur.pos += 1;
            let mut content: Vec<&str> = Vec::new();
            while let Some(l) = cur.peek() {
                cur.pos += 1;
                if fence_close(l, fence_len) {
                    break;
                }
                content.push(l);
            }
            out.push(Node::CodeBlock { info, text: content.join("\n") });
            continue;
        }

        if line.starts_with('>') && ctx.quote_depth < MAX_QUOTE_DEPTH {
            let mut inner_lines: Vec<&str> = Vec::new();
            while let Some(l) = cur.peek() {
                match l.strip_prefix('>') {
                    Some(rest) => {
                        inner_lines.push(rest.strip_prefix(' ').unwrap_or(rest));
                        cur.pos += 1;
                    }
                    None => break,
                }
            }
            let mut inner_ctx = ctx;
            inner_ctx.quote_depth += 1;
            let mut sub = Cursor { lines: inner_lines, pos: 0 };
            out.push(Node::Blockquote {
                children: parse_container(&mut sub, inner_ctx, None),
            });
            continue;
        }

        if line.trim_end_matches([' ', '\t']) == "---" {
            cur.pos += 1;
            out.push(Node::ThematicBreak);
            continue;
        }

        if ctx.list_depth < MAX_LIST_DEPTH {
            if let Some(m) = marker(line) {
                out.push(parse_list(cur, ctx, m));
                continue;
            }
        }

        // Paragraph: plain lines until blank / container end / block start.
        let mut plines = vec![line];
        cur.pos += 1;
        while let Some(l) = cur.peek() {
            if is_blank(l) || is_block_start(l, ctx) {
                break;
            }
            plines.push(l);
            cur.pos += 1;
        }
        if plines.len() == 1 {
            if let Some(target) = turbolink(plines[0]) {
                out.push(Node::Turbolink { target: target.to_string() });
                continue;
            }
        }
        out.push(Node::Paragraph {
            children: parse_inlines(&plines.join("\n")),
        });
    }
    out
}

fn invalid(reason: Reason) -> Node {
    Node::InvalidDirective { reason, children: Vec::new() }
}

fn is_blank(line: &str) -> bool {
    line.bytes().all(|b| b == b' ' || b == b'\t')
}

fn indent_of(line: &str) -> usize {
    line.bytes().take_while(|b| *b == b' ').count()
}

/// Would this line start a non-paragraph block in this context? (Used both
/// by the dispatcher and to end paragraphs: any block construct interrupts.)
fn is_block_start(line: &str, ctx: Ctx) -> bool {
    line.starts_with(":::")
        || line.starts_with("%%")
        || heading_line(line).is_some()
        || fence_open(line).is_some()
        || (line.starts_with('>') && ctx.quote_depth < MAX_QUOTE_DEPTH)
        || line.trim_end_matches([' ', '\t']) == "---"
        || (ctx.list_depth < MAX_LIST_DEPTH && marker(line).is_some())
}

fn heading_line(line: &str) -> Option<(u8, &str)> {
    let n = line.bytes().take_while(|b| *b == b'#').count();
    if (1..=8).contains(&n) && line.as_bytes().get(n) == Some(&b' ') {
        Some((n as u8, &line[n + 1..]))
    } else {
        None
    }
}

fn fence_open(line: &str) -> Option<(usize, Option<String>)> {
    let n = line.bytes().take_while(|b| *b == b'`').count();
    if n < 3 {
        return None;
    }
    let info = line[n..].trim_matches([' ', '\t']);
    Some((n, (!info.is_empty()).then(|| info.to_string())))
}

fn fence_close(line: &str, open_len: usize) -> bool {
    let t = line.trim_end_matches([' ', '\t']);
    t.len() >= open_len && t.bytes().all(|b| b == b'`')
}

struct Marker {
    indent: usize,
    ordered: bool,
    content_idx: usize,
}

fn marker(line: &str) -> Option<Marker> {
    let indent = indent_of(line);
    let rest = &line.as_bytes()[indent..];
    if rest.len() >= 2 && matches!(rest[0], b'-' | b'*' | b'+') && rest[1] == b' ' {
        return Some(Marker { indent, ordered: false, content_idx: indent + 2 });
    }
    let digits = rest.iter().take_while(|b| b.is_ascii_digit()).count();
    if digits > 0 && rest[digits..].starts_with(b". ") {
        return Some(Marker { indent, ordered: true, content_idx: indent + digits + 2 });
    }
    None
}

/// One list of one kind. A same-column marker of the other kind ends this
/// list (the dispatcher immediately starts the next one).
fn parse_list(cur: &mut Cursor, ctx: Ctx, first: Marker) -> Node {
    let ordered = first.ordered;
    let mut items: Vec<Node> = Vec::new();
    let mut buf: Vec<String> = vec![cur.peek().unwrap()[first.content_idx..].to_string()];
    cur.pos += 1;

    while let Some(line) = cur.peek() {
        if is_blank(line) {
            // The list continues past blanks only into indented content or a
            // same-kind column-0/1 marker; anything else ends it here.
            let mut j = cur.pos + 1;
            while j < cur.lines.len() && is_blank(cur.lines[j]) {
                j += 1;
            }
            let continues = match cur.lines.get(j) {
                None => false,
                Some(l) => {
                    indent_of(l) >= 2
                        || marker(l).is_some_and(|m| m.ordered == ordered && m.indent < 2)
                }
            };
            if !continues {
                break;
            }
            buf.push(String::new());
            cur.pos += 1;
            continue;
        }
        if indent_of(line) >= 2 {
            // Content (or a deeper marker) inside the current item; strip the
            // content column. Off-grid extra spaces ride along (floor rule).
            buf.push(line[2..].to_string());
            cur.pos += 1;
            continue;
        }
        // Column 0 or 1 (floors to 0).
        match marker(line) {
            Some(m) if m.ordered == ordered => {
                items.push(finish_item(std::mem::take(&mut buf), ctx));
                buf.push(line[m.content_idx..].to_string());
                cur.pos += 1;
            }
            _ => break, // column-0 block, or a kind switch
        }
    }
    items.push(finish_item(buf, ctx));
    Node::List { ordered, children: items }
}

fn finish_item(buf: Vec<String>, ctx: Ctx) -> Node {
    let mut inner_ctx = ctx;
    inner_ctx.list_depth += 1;
    let mut sub = Cursor {
        lines: buf.iter().map(|s| s.as_str()).collect(),
        pos: 0,
    };
    Node::ListItem {
        children: parse_container(&mut sub, inner_ctx, None),
    }
}

enum DirLine {
    Open { name: String, attrs_src: String, leaf: bool },
    Close(Option<String>),
    Bad(Reason),
}

fn classify_directive(line: &str) -> DirLine {
    let rest = line[3..].trim_end_matches([' ', '\t']);
    if rest.is_empty() {
        return DirLine::Close(None);
    }
    if rest.starts_with(' ') || rest.starts_with('\t') {
        // `::: name` - a named close.
        let name = rest.trim_matches([' ', '\t']);
        return if is_name(name) {
            DirLine::Close(Some(name.to_string()))
        } else {
            DirLine::Bad(Reason::BadName)
        };
    }
    // Open line. The leaf closer token is stripped before attribute parsing.
    let (body, leaf) = match rest.strip_suffix(":::") {
        Some(b) => (b, true),
        None => (rest, false),
    };
    let nlen = name_len(body);
    if nlen == 0 {
        return DirLine::Bad(Reason::BadName);
    }
    let after = &body[nlen..];
    if !(after.is_empty() || after.starts_with(' ') || after.starts_with('\t')) {
        return DirLine::Bad(Reason::BadName);
    }
    DirLine::Open {
        name: body[..nlen].to_string(),
        attrs_src: after.to_string(),
        leaf,
    }
}

/// A paragraph that is exactly one authority-form absolute URI is a
/// turbolink. Every bare word is a valid relative URI reference, so the
/// sugar demands `scheme://`; everything else uses `:::turbolink`.
fn turbolink(line: &str) -> Option<&str> {
    let t = line.trim_matches([' ', '\t']);
    if t.is_empty() || t.len() > MAX_TARGET_BYTES {
        return None;
    }
    if t.contains(' ') || t.contains('\t') {
        return None;
    }
    let b = t.as_bytes();
    if !b[0].is_ascii_alphabetic() {
        return None;
    }
    let scheme_len = b
        .iter()
        .take_while(|c| c.is_ascii_alphanumeric() || matches!(c, b'+' | b'-' | b'.'))
        .count();
    let after = &t[scheme_len..];
    if after.starts_with("://") && after.len() > 3 {
        Some(t)
    } else {
        None
    }
}
